sbi/inference/potentials/posterior_based_potential.py

```diff
@@ -100,29 +97,84 @@ def __call__(self, theta: Tensor, track_gradients: bool = True) -> Tensor:
                 "No observed data x_o is available. Please reinitialize \
                 the potential or manually set self._x_o."
             )
+        # Force probability to be zero outside prior support.
+
+        in_prior_support = within_support(self.prior, theta)
+        if self.x_is_iid:
+            # Calculate posterior probabilities over trials and in one batch.
+            posterior_log_prob = _log_posteriors_over_trials(
+                x=self._x_o,
+                theta=theta.to(self.device),
+                estimator=self.posterior_estimator,
+                track_gradients=track_gradients,
+            )
+        else:
+            theta_batch_size = theta.shape[0]
+            x_batch_size = self._x_o.shape[0]
+            assert (
+                theta_batch_size == x_batch_size
+            ), f"Batch size mismatch: {theta_batch_size} and {x_batch_size}.\
+                When performing batched sampling for multiple `x`, the batch size of\
+                `theta` must match the batch size of `x`."
+
+            theta = theta.unsqueeze(0)
+            with torch.set_grad_enabled(track_gradients):
+                posterior_log_prob = self.posterior_estimator.log_prob(
+                    theta, condition=self._x_o
+                )
+                posterior_log_prob = torch.where(
+                    in_prior_support,
+                    posterior_log_prob,
+                    torch.tensor(
+                        float("-inf"), dtype=torch.float32, device=self.device
+                    ),
+                )
+        return posterior_log_prob
 
-        theta = ensure_theta_batched(torch.as_tensor(theta)).to(self.device)
 
-        with torch.set_grad_enabled(track_gradients):
-            # Force probability to be zero outside prior support.
-            in_prior_support = within_support(self.prior, theta)
+def _log_posteriors_over_trials(
+    x: Tensor,
+    theta: Tensor,
+    estimator: ConditionalDensityEstimator,
+    track_gradients: bool = False,
+) -> Tensor:
```

Comment:
I am confused about this function. In my head, when doing posterior estimation, we would usually have a permutation invariant embedding net for `x` that takes care of embedding all the trials, before actually evaluating the density estimator.

Reply:
How does this function fit into that model?